### DISCUSSION TITLE
Can cue validate number in a string?

### DISCUSSION THREAD

**Discussion by beviz** [cue-lang/cue#1803] [Q&A] [answered]
for example:

```YAML
percentages:
  - '10%'
  - '20%'
  - '30%' // sum
```

I want to validate that the third value is the sum of the rest. Does cue support this?
I'm considering using CUE in our project. This is the only validation I've not found.

**Comment by Lyoness** [comment]
Thanks for writing in and sorry we missed this. Hope you found an answer. 

For future reference: The CUE [Github Discussions space](https://github.com/cue-lang/cue/discussions) is the place you can ask these types of questions, for the benefit of all who might have similar questions in future (and also to help improve documenation). 

 You can also chat about what you're trying to do and how CUE might help in the [community slack](https://join.slack.com/t/cuelang/shared_invite/enQtNzQwODc3NzYzNTA0LTAxNWQwZGU2YWFiOWFiOWQ4MjVjNGQ2ZTNlMmIxODc4MDVjMDg5YmIyOTMyMjQ2MTkzMTU5ZjA1OGE0OGE1NmE).

**Comment by Lyoness** [comment]
@beviz  I've converted this issue to a discussion. Better late than never?

Also will help future similar questions that may come (and improve docs).

**Comment by slewiskelly** [comment, marked as answer]
Numbers in a string can be validated using regex, so ensuring the string is a valid percentage between 0-100%:

```cue
percentages: [...string & =~"^([0-9]%$|^[1-9][0-9]%$|^100%)$"]
```

However, this is not quite what you need in validating that the last element is the sum of the remaining elements.

I can't think of an extremely straightforward way, but it is possible:

```cue
import (
	"list"
	"strconv"
	"strings"
)

percentages: [
	"10%",
	"20%",
	"30%",
]

_percentages: [...int & >=0 & <=100] & [ for p in percentages {strconv.Atoi(strings.Replace(p, "%", "", -1))}]

_lastIsSumOfRest: true & list.Sum(_percentages)-_percentages[len(_percentages)-1] == _percentages[len(_percentages)-1]
```

**Reply by beviz** [reply]
Thank you @slewiskelly, it could be a workaround, though it's a bit ugly. Marked as answer.
